### PR TITLE
fix: exclude env vars from build pack if passed in

### LINKED ISSUE #6313: create effective pipeline fails if passing in DOCKER_REGISTRY
if we pass in the env var DOCKER_REGISTRY to creating the effective pipeline we've seen failures where the pod gets created with an env var of the form:

```yaml
    - name: DOCKER_REGISTRY
      value: gcr.io
      valueFrom:
        configMapKeyRef:
          key: docker.registry
          name: jenkins-x-docker-registry
```

which is invalid; either `value` or `valueFrom` should be used.

It looks like this is coming from logic of taking the `DOCKER_REGISTRY` from the pod templates and not detecting its being passed in to the meta pipeline client as a custom environment variable so should be discarded from the build pack. i.e. it should just generate this...

```yaml
    - name: DOCKER_REGISTRY
      value: gcr.io
```

### PR DESCRIPTION
**Pull request by jstrachan**
if we pass in custom env vars when creating an effective pipeline we should remove any inherited env vars of the same name from the build pack to avoid creating invalid combined env var definitions in the pods which tekton + k8s cannot instantiate

fixes #6313

Signed-off-by: James Strachan <james.strachan@gmail.com>

### CODE REVIEW

File: pkg/cmd/step/syntax/step_syntax_effective.go [hunk 1]
```diff
@@ -420,6 +420,21 @@ func (o *StepSyntaxEffectiveOptions) createPipelineForKind(kind string, lifecycl
...
 		}
+
+		// lets remove any environment variables defined in the container options from build packs/pod templates
+		// which are explicitly passed in to avoid possibly adding a value and valueFrom expression in the
+		// generated pod's container env vars
+		co := parsed.Options.ContainerOptions
+		for _, ev := range pipelineConfig.Env {
+			name := ev.Name
+			for i, e := range co.Env {
+				if e.Name == name {
```
Comment (hferentschik) [inline pkg/cmd/step/syntax/step_syntax_effective.go]: So this is the actual removal from container options, right?

File: pkg/cmd/step/syntax/step_syntax_effective.go [hunk 2]
```diff
@@ -420,6 +420,21 @@ func (o *StepSyntaxEffectiveOptions) createPipelineForKind(kind string, lifecycl
...
+					co.Env = append(co.Env[0:i], co.Env[i+1:]...)
+					break
+				}
+			}
+		}
+		co.Env = RemoveDuplicateEnvVars(co.Env)
```
Comment (hferentschik) [inline pkg/cmd/step/syntax/step_syntax_effective.go]: Is this an additional step? This is has nothing to do with the removal between duplicated in container options and pipeline env, right?

File: pkg/cmd/step/syntax/step_syntax_effective.go [hunk 3]
```diff
@@ -451,6 +466,26 @@ func (o *StepSyntaxEffectiveOptions) createPipelineForKind(kind string, lifecycl
 	return parsed, nil
 }
 
+// RemoveDuplicateEnvVars removes any duplicate env var definitions
+func RemoveDuplicateEnvVars(env []corev1.EnvVar) []corev1.EnvVar {
+	counts := map[string]int{}
+	for _, e := range env {
+		counts[e.Name] = counts[e.Name] + 1
+	}
+	for name, c := range counts {
```
Comment (hferentschik) [inline pkg/cmd/step/syntax/step_syntax_effective.go]: We don't care which env variable is left in the end as long as there are no duplicated?

File: pkg/cmd/step/syntax/step_syntax_effective.go (hunk 2 above)
Comment (abayer) [reply pkg/cmd/step/syntax/step_syntax_effective.go]: Yeah, I'm very confused at this point. =)

Comment (hferentschik) [reply pkg/cmd/step/syntax/step_syntax_effective.go]: Still not sure what the intend of ` RemoveDuplicateEnvVars` is as that stage.

**Comment by abayer** [comment]
@jstrachan - so lemme try to make sure I understand what this is doing - it's meant to have anything specified in `pipelineConfig.Env` override any env vars specified in pod templates or the build pack? It looks to me like it just gets env vars from pod templates, since it's just looking at the container options, not the stage/step env. Is that right?

**Comment by abayer** [comment]
You might want to take a look at `combineEnvVars` in `step_syntax_effective` - it's currently specific to using `o.CustomEnvs` to override `projectConfig.Env`, but its cousin `scopedEnv` in `tekton/syntax/pipeline.go` could get moved to utils somewhere and be used as a general purpose "take this first list of env vars and this existing list of env vars and add the first to the existing, overriding anything that was there" function.

**Comment by abayer** [comment]
(and yeah, any `Env` other than in `ContainerOptions` will be properly overridden at time of translation to Tekton CRDs - `ContainerOptions` is the only place where that doesn't already happen)

File: pkg/cmd/step/syntax/step_syntax_effective.go (hunk 3 above)
Comment (jstrachan) [reply pkg/cmd/step/syntax/step_syntax_effective.go]: fwiw the existing test cases were generating duplicate name + values which is why I added this code. I could always just leave the duplicates there if you'd rather?

File: pkg/cmd/step/syntax/step_syntax_effective.go (hunk 2 above)
Comment (jstrachan) [reply pkg/cmd/step/syntax/step_syntax_effective.go]: I'm confused too - removing duplicate values resulted in duplicates in the test case

**Comment by jstrachan** [comment]
@abayer @hferentschik thanks for the feedback - am gonna try a much simpler + way less confusing fix instead https://github.com/jenkins-x/jx/pull/6316 gonna close this one for now

**Comment by hferentschik** [comment]
> thanks for the feedback - am gonna try a much simpler + way less confusing fix instead #6316 gonna close this one for now

awesome thanks

**Pull request closed without merging**